### ISSUE TITLE
NETCore 2.0 Preview1: nvalidOperationException: Cannot resolve scoped service 'FluentValidation.IValidatorFactory' from root provider.

### ISSUE THREAD

**Issue by jstynder** [JeremySkinner/FluentValidation#474]
Hi Jeremy,
Create a new ASP.NET Core 2.0 Preview1 application , and running into this error 

**InvalidOperationException: Cannot resolve scoped service 'FluentValidation.IValidatorFactory' from root provider.**

Startup ConfigureServices 
services.AddMvc().AddFluentValidation(); 

See Stack Trace below:

Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteValidator.ValidateResolution(Type serviceType, ServiceProvider serviceProvider)
Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetService<T>(IServiceProvider provider)
FluentValidation.AspNetCore.FluentValidationMvcExtensions+<>c__DisplayClass1_1.<RegisterServices>b__2(IServiceProvider s)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitFactoryService(FactoryService factoryService, ServiceProvider provider)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitTransient(TransientCallSite transientCallSite, ServiceProvider provider)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitClosedIEnumerable(ClosedIEnumerableCallSite closedIEnumerableCallSite, ServiceProvider provider)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitTransient(TransientCallSite transientCallSite, ServiceProvider provider)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(SingletonCallSite singletonCallSite, ServiceProvider provider)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitConstructor(ConstructorCallSite constructorCallSite, ServiceProvider provider)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitScoped(ScopedCallSite scopedCallSite, ServiceProvider provider)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitSingleton(SingletonCallSite singletonCallSite, ServiceProvider provider)
Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor.VisitCallSite(IServiceCallSite callSite, TArgument argument)
Microsoft.Extensions.DependencyInjection.ServiceProvider+<>c__DisplayClass17_0.<RealizeService>b__0(ServiceProvider provider)
Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType)
Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService<T>(IServiceProvider provider)
Microsoft.AspNetCore.Mvc.ViewResult+<ExecuteResultAsync>d__26.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Mvc.Core.Internal.ResourceInvoker+<InvokeResultAsync>d__15.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker+<InvokeNextResultFilterAsync>d__19.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Rethrow(ResultExecutedContext context)
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Internal.ControllerActionInvoker+<InvokeInnerFilterAsync>d__21.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Mvc.Core.Internal.ResourceInvoker+<InvokeNextResourceFilter>d__18.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
Microsoft.AspNetCore.Mvc.Core.Internal.ResourceInvoker.Rethrow(ResourceExecutedContext context)
Microsoft.AspNetCore.Mvc.Core.Internal.ResourceInvoker.Next(ref State next, ref Scope scope, ref object state, ref bool isCompleted)
Microsoft.AspNetCore.Mvc.Core.Internal.ResourceInvoker+<InvokeFilterPipelineAsync>d__13.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Mvc.Core.Internal.ResourceInvoker+<InvokeAsync>d__11.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Builder.RouterMiddleware+<Invoke>d__4.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Authentication.AuthenticationMiddleware+<Invoke>d__6.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Diagnostics.Identity.Service.DeveloperCertificateMiddleware+<InvokeAsync>d__7.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.MigrationsEndPointMiddleware+<Invoke>d__5.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseErrorPageMiddleware+<Invoke>d__6.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
Microsoft.AspNetCore.Diagnostics.EntityFrameworkCore.DatabaseErrorPageMiddleware+<Invoke>d__6.MoveNext()
System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware+<Invoke>d__7.MoveNext()

**Comment by JeremySkinner** [comment]
Thanks, but there's no support for .net core 2 at present. I'll look at adding it when it's closer to RTM. If there's one thing I learned from the .net core 1.x development it's to stay well away from it until it's stable.

**Comment by moattarwork** [comment]
This is happening in SDK 1.0.4 too.

**Comment by JeremySkinner** [comment]
@moattarwork Are you on Microsoft.AspNetCore.Mvc 1.0.4 too? If so you need to be on at least 1.1.2 for compatibility with FluentValidation (and FV 7.0.2)

**Comment by pmezykowski** [comment]
First, I would like to agree with @JeremySkinner  that fixing it on current stage is probably pointless, but @jstynder if you still need quick and dirty workaround, you can add the factory manually as a transient:
```C#
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddMvc()
                .AddFluentValidation(c => c.RegisterValidatorsFromAssemblyContaining<Startup>());
            services.AddTransient<IValidatorFactory, ServiceProviderValidatorFactory>();
        }
```
I think I don't have to add that you should not use it in production code:)

Cheers:)

**Comment by thepenumbra** [comment]
Any update on this? With the release of stable version of .net core 2.0 this error is still exist.

**Comment by JeremySkinner** [comment]
I actually tried this with the final release of .net core 2 over the weekend and couldn't reproduce the problem. Everything was resolved correctly from the container. 

Could you put together a small sample that reproduces the issue? I'll then be able to investigate further. Thanks.

**Comment by thepenumbra** [comment]
Thank you for your quick response. I attach a sample project that reproduce this issue.
[Test3WebApplication.zip](https://github.com/JeremySkinner/FluentValidation/files/1241161/Test3WebApplication.zip)

**Comment by JeremySkinner** [comment]
@thepenumbra 
Ok I had a quick look at this. The problem isn't a bug in FluentValidation, but it's an issue with your configuration. The IValidatorFactory is registered as request-scoped, but you're depending on it within a singleton-scoped service:

```
 services.AddSingleton<ISingletonService, SingletonService>();
 services.AddSingleton<Func<IValidatorFactory>>((s) => () => s.GetService<IValidatorFactory>());
```

That's not going to work - you can't have a singleton service depend on a non-singleton service. Even using a Func in this way won't work, as the service provider inside the func is still the root service provider. You need to restructure your services so that you dont have a singleton depending on something that's request-scoped.

**Comment by JeremySkinner** [comment]
@thepenumbra 
Ok I had a quick look at this. The problem isn't a bug in FluentValidation, but it's an issue with your configuration. The IValidatorFactory is registered as request-scoped, but you're depending on it within a singleton-scoped service:

```
 services.AddSingleton<ISingletonService, SingletonService>();
 services.AddSingleton<Func<IValidatorFactory>>((s) => () => s.GetService<IValidatorFactory>());
```

That's not going to work - you can't have a singleton service depend on a non-singleton service. Even using a Func in this way won't work, as the service provider inside the func is still the root service provider. You need to restructure your services so that you dont have a singleton depending on something that's request-scoped.

**Comment by thepenumbra** [comment]
OK I got it. Thanks, but is it a good idea to be able to configure how IValidatorFactory instantiate? Probably an option in FluentValidationMvcConfiguration?

**Comment by JeremySkinner** [comment]
There's no need for it to be in FluentValidationMvcConfiguration - you can do it yourself by just calling into the service provider directly (Eg by calling `services.AddTransient<IValidatorFactory, ServiceProviderValidatorFactory>()` after the call to AddFluentValidation)

**Comment by thepenumbra** [comment]
If i have a service that depends on a repository that needs (or good) to be singleton (e.g. a nosql db repository), I can't define the service as scoped or transient. I think calling services.AddTransient<IValidatorFactory, ServiceProviderValidatorFactory>() twice is just a workaround.

**Comment by juchom** [comment]
After upgrading my website from net core 1.1 I have the same error with asp.net core 2.0.

I tried creating a new web site from the template in net core 2.0
I added the FluentValidation.AspNetCore
I've changed the line from:
`services.AddMvc();`
to
`services.AddMvc().AddFluentValidation();`

And the website throws on startup.

**Comment by JeremySkinner** [comment]
@juchom you need to ensure that you don't have any singleton scoped services that depend on IValidatorFactory (which is request scoped).

**Comment by juchom** [comment]
Thanks for the quick reply,

I have this error even with the basic web site template of VS2017 15.3.

Here is the basic Startup.cs

```
    public class Startup
    {
        public Startup(IConfiguration configuration)
        {
            Configuration = configuration;
        }

        public IConfiguration Configuration { get; }

        // This method gets called by the runtime. Use this method to add services to the container.
        public void ConfigureServices(IServiceCollection services)
        {
            services.AddMvc().AddFluentValidation();
        }

        // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
        public void Configure(IApplicationBuilder app, IHostingEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
                app.UseBrowserLink();
            }
            else
            {
                app.UseExceptionHandler("/Error");
            }

            app.UseStaticFiles();

            app.UseMvc(routes =>
            {
                routes.MapRoute(
                    name: "default",
                    template: "{controller}/{action=Index}/{id?}");
            });
        }
    }
```

**Comment by JeremySkinner** [comment]
I have the same and it works fine - which build of FluentValidation are you running?

**Comment by juchom** [comment]
Here is the csproj

```
<Project Sdk="Microsoft.NET.Sdk.Web">
  <PropertyGroup>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="FluentValidation.AspNetCore" Version="7.1.1" />
    <PackageReference Include="Microsoft.AspNetCore.All" Version="2.0.0" />
  </ItemGroup>
  <ItemGroup>
    <DotNetCliToolReference Include="Microsoft.VisualStudio.Web.CodeGeneration.Tools" Version="2.0.0" />
  </ItemGroup>
</Project>

```

**Comment by JeremySkinner** [comment]
Does it make a different if you try it with FluentValidation.AspNetCore 7.2-beta1 ?

**Comment by juchom** [comment]
Same error, do you want a zip of the project?

**Comment by JeremySkinner** [comment]
Yes please, that'd be great.

**Comment by juchom** [comment]
Here it is. 

[WebApplication1.zip](https://github.com/JeremySkinner/FluentValidation/files/1242498/WebApplication1.zip)

Let me know, how I can assist you. Thanks a lot!

**Comment by JeremySkinner** [comment]
Thanks! I'll try and take a look at this later today.

**Comment by JeremySkinner** [comment]
Ok, it seems to be something related to Razor. Using your sample project, the error occurs when navigating to the default Index razor page, but does not occur when creating a simple controller action.

**Comment by JeremySkinner** [comment]
TL;DR - Fixed and pushed out 7.2-beta2 with the fix. Please could you all try updating to this build from nuget, and let me know if you run into any issues? 

**Long explanation:**

The issue is being caused by this line:

https://github.com/JeremySkinner/FluentValidation/blob/7604500a21e7a911fb6576a3985ce0cae9640c0c/src/FluentValidation.AspNetCore/FluentValidationMvcExtensions.cs#L86

...as it's related to ViewOptionsSetup, it only gets invoked when a Razor page is executed which is why I wasn't seeing it in my testing, as I wasn't executing any views. 

The service here is being registered as transient and is used to enable FV's clientside integration, so depends on the request-scoped IValidatorFactory. 

In theory, this should be fine - having a transient service depend on a request-scoped object is perfectly fine from a lifecycle point of view, what's not immediately apparent is that higher up in the dependency chain, `Microsoft.Extensions.Options.IOptions<Microsoft.AspNetCore.Mvc.MvcViewOptions>` is actually a singleton! This means that even though my registration was transient, it's actually using the root (singleton) container to do the resolution. :man_facepalming: 

I guess the solution is to make IValidatorFactory always be registered as transient. I don't really like this as it means there can be multiple IValidatorFactory instances floating around, but I can't think of another solution.

**Comment by juchom** [comment]
Works for me!

Thanks a lot Jeremy!

**Comment by JeremySkinner** [comment]
The reason I wasn't able to reproduce this is due to the environment- the error only occurs if the environment is set to "Development". If you run in "Production" mode, the issue does not occur. :man_shrugging:

**Comment by thepenumbra** [comment]
Actually this error relates to https://github.com/aspnet/MetaPackages/blob/586d641934e997045ed98b1978684a7b0c607b39/src/Microsoft.AspNetCore/WebHost.cs#L189
If You call UseDefaultServiceProvider in Program and pass `(s) => s.ValidateScopes = false` to it, you wouldn't get this error even you were in "Development" environment.

**Comment by JeremySkinner** [comment]
Thanks, that's helpful. I've modified the integration tests for enable scope validation, so I'll catch this if it happens elsewhere in the codebase. 

https://github.com/JeremySkinner/FluentValidation/blob/4211590371a97a2d9bd78af87980be9025104a69/src/FluentValidation.Tests.Mvc6.dotnet/MvcIntegrationTests.cs#L18-L25

(UseDefaultServiceProvider is only available on .net core 2, but not .net core 1, and the integration tests are shared across both, hence the conditional statement)

**Comment by thepenumbra** [comment]
Alternatively in your test project, you could use `UseEnvironment("Development")` in WebHostBuilder initialization to override the default "Production" environment. It exists in the .NET Core 1.x and therefore no need for using preprocessor directive.

**Comment by JeremySkinner** [comment]
I actually tried that, but UseEnvironment didn't seem to set options.ValidateScopes = true when inside the integration tests. I was probably missing something, but didn't have the time to dig into it too deeply.

**Comment by thepenumbra** [comment]
I think the problem is using `new WebHostBuilder()` instead of `WebHost.CreateDefaultBuilder()`.

**Comment by dodyg** [comment]
I am using ```7.2-beta2``` and .NET Core 2.0. This configuration does not work for me

```
            services
                .AddMvc()             
     .AddViewLocalization(Microsoft.AspNetCore.Mvc.Razor.LanguageViewLocationExpanderFormat.Suffix)
                .AddFluentValidation( fv =>
                {
                    fv.RegisterValidatorsFromAssemblyContaining<RuntimeConfig>();
                });
```

```RuntimeConfig``` is just a class in another assembly that contains the validators. Is there anything else I am missing? This is one of my validator

```
    public class MeasuringUnitEditValidator : AbstractValidator<edit.MeasuringUnitEdit>
    {
        public MeasuringUnitEditValidator(IStringLocalizer<Core.Resources.Global> _global)
        {
            RuleFor(x => x.Name).NotEmpty().WithMessage(_global["NameIsRequired"]);
        }
    }
```

**Comment by JeremySkinner** [comment]
@dodyg Please could you open a separate issue. This issue is related to IoC/dependency scoping. I'll also need you to provide more details of what's not working. Thanks.

**Comment by dodyg** [comment]
https://github.com/JeremySkinner/FluentValidation/issues/551

**Comment by JeremySkinner** [comment]
@thepenumbra Thanks, but that WebHost class was also only added in netcore 2, so would still require the conditional compilation. Will leave it as-is as it's working now. Thanks for your help with this - it's much appreciated.
